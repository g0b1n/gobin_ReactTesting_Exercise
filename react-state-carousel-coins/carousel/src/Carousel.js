import { useState } from "react";
import "./Carousel.css";
import Card from "./Card";


/** Carousel: displays images and arrows to navigate through them
 * 
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 * 
 * State:
 * - currCardIdx: integer for current card index
 * 
 * App --> Carousel --> Card
 */
 function Carousel({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0);

  const currCard = photos[currCardIdx];
  const total = photos.length;

  //Increments currCardIdx state by 1
  function goForward() {
    setCurrCardIdx(currCardIdx + 1);
  }


  // FIX BUG
  // Decrements currCardIdx state by -1
  function goBackward() {
    setCurrCardIdx(currCardIdx - 1)
  }

  


  return (
  //   <div className="Carousel">
  //     <h1>{title}</h1>
  //     <div className="Carousel-main">
  //       <i
  //         className="bi-arrow-left-circle"
  //         onClick={goBackward}
  //       />
  //       <Card
  //         caption={currCard.caption}
  //         src={currCard.src}
  //         currNum={currCardIdx + 1}
  //         totalNum={total}
  //       />
  //       <i
  //         className="bi bi-arrow-right-circle"
  //         onClick={goForward}
  //       />
  //     </div>
  //   </div>

  // Updated to hide/show the arrows based on the currCardIdx of imgs
  <div className="Carousel">
    <h1>{ title }</h1>
    <div className="Carousel-main">
      {currCardIdx > 0 && (
        <i 
          className="bi bi-arrow-left-circle"
          data-testid="left-arrow"
          onClick={goBackward}
          />
      )}
      <Card
        caption={currCard.caption}
        src={currCard.src}
        currNum={currCardIdx + 1}
        totalNum={total}
      />
      {currCardIdx < total - 1 && (
        <i
          className="bi bi-arrow-right-circle"
          data-testid="right-arrow"
          onClick={goForward}
        />
      )}
    </div>
  </div>
  )
}

export default Carousel;
