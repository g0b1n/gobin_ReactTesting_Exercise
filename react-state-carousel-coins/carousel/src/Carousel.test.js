import { render, fireEvent } from "@testing-library/react";
import Carousel from "./Carousel";
import TEST_IMAGES from "./_testCommon.js";

it("works when you click on the right arrow", function() {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );
  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument();

  // move forward in the carousel
  const rightArrow = container.querySelector(".bi-arrow-right-circle");
  fireEvent.click(rightArrow);

  // move backward in the carousel
  const leftArrow = container.querySelector(".bi-arrow-left-circle");
  fireEvent.click(leftArrow);

  // expect the second image to show, but not the first
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument();
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument();
});

// Smoke Test
it("should render carousal", () => {
  const dummyImg = [
    { src: "hppts://bestImage.com/img1.jpg", caption: "img1"},
    { src: "hppts://bestImage.com/img2.jpg", caption: "img2"}
  ]
  render(<Carousel photos={dummyImg} title="testing"/>)
})

// Snapshot Test
it("should match the snapshot", () => {
  const dummyImg = [
    { src: "hppts://bestImage.com/img1.jpg", caption: "img1"},
    { src: "hppts://bestImage.com/img2.jpg", caption: "img2"}
  ]
  const { asFragment } = render(<Carousel photos={dummyImg} title="testing" />);
  expect(asFragment()).toMatchSnapshot()
})


// Testing if the Arrows are hidden or not

// Test if left arrow is hidden when on 1st img
it("hides left arrow when on the first img", function() {
  const {queryByTestId} = render(
    <Carousel
    photos={TEST_IMAGES}
    title="testing imgs"
    />
  );

  // check if the left arrow is hidden from DOM
  const leftArrow = queryByTestId('left-arrow');
  expect(leftArrow).not.toBeInTheDocument();
});

// Test if right arrow is hideen when on last img
it("hides right arrow when on the last image", function() {
  const { container, queryByTestId } = render(
    <Carousel
      photos={TEST_IMAGES}
      title="images for testing"
    />
  );

  // move to last img
  const rightArrow = container.querySelector(".bi-arrow-right-circle");
  for (let i = 0; i < TEST_IMAGES.length - 1; i++) {
    fireEvent.click(rightArrow);
  }

  // check if the right arrow is hidden when on the last img
  const lastImageRightArrow = queryByTestId("right-arrow");
  expect(lastImageRightArrow).not.toBeInTheDocument();
});