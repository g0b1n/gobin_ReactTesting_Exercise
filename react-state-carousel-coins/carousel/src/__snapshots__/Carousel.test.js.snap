// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      testing
    </h1>
    <div
      class="Carousel-main"
    >
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          img1
        </h4>
        <img
          alt="img1"
          class="Card-image"
          src="hppts://bestImage.com/img1.jpg"
        />
        <small
          class="Card-small"
        >
          Image 1 of 2.
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
        data-testid="right-arrow"
      />
    </div>
  </div>
</DocumentFragment>
`;
